leftwm-core/src/handlers/command_handler.rs

```diff
@@ -557,28 +572,12 @@ fn toggle_floating(state: &mut State) -> Option<bool> {
     }
 }
 
-fn move_focus_common_vars<F>(func: F, state: &mut State, val: i32) -> Option<bool>
-where
-    F: Fn(&mut State, i32, WindowHandle, Option<Layout>, Vec<Window>) -> Option<bool>,
-{
-    let handle = state.focus_manager.window(&state.windows)?.handle;
-    let tag_id = state.focus_manager.tag(0)?;
-    let tag = state.tags.get(tag_id)?;
-    let (tags, layout) = (vec![tag_id], Some(tag.layout));
-
-    let for_active_workspace =
-        |x: &Window| -> bool { helpers::intersect(&tags, &x.tags) && !x.is_unmanaged() };
-
-    let to_reorder = helpers::vec_extract(&mut state.windows, for_active_workspace);
-    func(state, val, handle, layout, to_reorder)
-}
-
 fn move_window_change(
     state: &mut State,
-    val: i32,
     mut handle: WindowHandle,
     layout: Option<Layout>,
     mut to_reorder: Vec<Window>,
+    val: i32,
```

Comment:
Just curious: Is this reordering meaningful and I miss the implications, or is it simply aesthetic?

Reply:
The ordering matters.

Some of the arguments are created by the wrapper function/macro `move_focus_common_vars`. With the function, the argument `val` was arbitrary and sometimes not used. So that one goes to `$arg` in the macro (which is multiple! very cool)

But the leading argument must be common for all the functions. That why I moved `val` to the end.

An alternative to the macro would be to use a generic type argument... tbh I think macros are nicer in this case because it allows having multiple custom arguments.

I'm also not super happy with this wrapper function (`move_focus_common_vars`)... but I'm not sure how to change it and it's not really my priority so... :woman_shrugging: good enough for now I guess.

Reply:
Ohh, the macro brought that, thanks for the elaborate explanation!

Iirc the `..._common_vars` were part of some performance optimisations with significant impact by @AethanFoot

Same as with macros mechanics these are hard for me to grasp with my layman greenhorn knowledge though :joy:

Reply:
Performance? Oh that explains it!

Well now it's performant and flexible xD

Reply:
@VuiMuich you can see in this code how the arguments are dynamically added depending on how the macro is called:

```rust
$func($state, handle, layout, to_reorder, $($arg),*)
```

- `$func` is your inner function.
- `$state` the state provided in argument of the macro
- `handle`, `layout`, `to_recorder` are computed in that macro
- `$( ... ),*` is like `$( ... )*`: expanding as many `$arg` as provided to the macro call, but the `,` indicates they will be joined with a `,` in between

This would work the same:

```rust
$func($state, handle, layout, to_reorder $(, $arg)*)
```

(It's slightly better because the `,` after `to_reorder` is only inserted if there are arguments. But that just compile-time cost so it really does not matter at all.)